tests/test_h2_client.c

```diff
@@ -1831,56 +2284,86 @@ TEST_CASE(h2_client_conn_err_initial_window_size_cause_window_exceed_max) {
+    ASSERT_TRUE(aws_byte_buf_eq_c_str(&sent_data_frame->data, "h"));
+    /* fake peer sends complete response */
+    struct aws_http_header response_headers_src[] = {
+        DEFINE_HEADER(":status", "404"),
+    };
+    /* stop stalling the input stream */
+    aws_input_stream_tester_set_max_bytes_per_read(request_body, 5);
+    size_t frames_count = h2_decode_tester_frame_count(&s_tester.peer.decode);
 
-    /* fake peer sends RST_STREAM */
-    struct aws_h2_frame *rst_stream = aws_h2_frame_new_rst_stream(allocator, stream_id, AWS_H2_ERR_HTTP_1_1_REQUIRED);
-    ASSERT_SUCCESS(h2_fake_peer_send_frame(&s_tester.peer, rst_stream));
+    struct aws_http_headers *response_headers = aws_http_headers_new(allocator);
+    aws_http_headers_add_array(response_headers, response_headers_src, AWS_ARRAY_SIZE(response_headers_src));
 
-    /* validate that stream completed with error */
+    struct aws_h2_frame *response_frame =
+        aws_h2_frame_new_headers(allocator, stream_id, response_headers, true /*end_stream*/, 0, NULL);
+    ASSERT_SUCCESS(h2_fake_peer_send_frame(&s_tester.peer, response_frame));
+
+    /* No rst stream sent, we wait until the client finish sending body */
+    /* validate the client request completes successfully */
     testing_channel_drain_queued_tasks(&s_tester.testing_channel);
     ASSERT_TRUE(stream_tester.complete);
-    ASSERT_INT_EQUALS(AWS_ERROR_HTTP_RST_STREAM_RECEIVED, stream_tester.on_complete_error_code);
-
-    /* a stream error should not affect the connection */
-    ASSERT_TRUE(aws_http_connection_is_open(s_tester.connection));
+    ASSERT_INT_EQUALS(AWS_ERROR_SUCCESS, stream_tester.on_complete_error_code);
+    ASSERT_INT_EQUALS(404, stream_tester.response_status);
 
-    /* validate that stream did NOT send RST_STREAM */
+    /* Check the rest of the body received by peer */
     ASSERT_SUCCESS(h2_fake_peer_decode_messages_from_testing_channel(&s_tester.peer));
-    ASSERT_NULL(h2_decode_tester_find_frame(&s_tester.peer.decode, AWS_H2_FRAME_T_RST_STREAM, 0, NULL));
+    struct h2_decoded_frame *rest_data_frame = h2_decode_tester_find_frame(
+        &s_tester.peer.decode, AWS_H2_FRAME_T_DATA, frames_count /*search_start_idx*/, NULL);
+    ASSERT_TRUE(rest_data_frame->end_stream);
+    ASSERT_TRUE(aws_byte_buf_eq_c_str(&rest_data_frame->data, "ello"));
 
     /* clean up */
-    aws_http_message_release(request);
+    aws_http_headers_release(response_headers);
     client_stream_tester_clean_up(&stream_tester);
+    aws_http_message_release(request);
+    aws_input_stream_destroy(request_body);
     return s_tester_clean_up();
 }
 
 /* A server MAY request that the client abort transmission of a request without error by sending a
  * RST_STREAM with an error code of NO_ERROR after sending a complete response. */
-TEST_CASE(h2_client_stream_receive_rst_stream_after_complete_response_ok) {
+TEST_CASE(h2_client_stream_receive_end_stream_and_rst_before_done_sending) {
```

Comment:
Actually an old test, just rename it, and made one more check, you can skip when you review.